airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination/record_buffer/SerializedRecordBufferingStrategy.java

```diff
@@ -0,0 +1,126 @@
+import io.airbyte.commons.functional.CheckedBiConsumer;
+import io.airbyte.commons.functional.CheckedBiFunction;
+import io.airbyte.commons.string.Strings;
+import io.airbyte.integrations.base.AirbyteStreamNameNamespacePair;
+import io.airbyte.protocol.models.AirbyteMessage;
+import io.airbyte.protocol.models.ConfiguredAirbyteCatalog;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import org.apache.commons.io.FileUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class SerializedRecordBufferingStrategy implements RecordBufferingStrategy {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(SerializedRecordBufferingStrategy.class);
+
+  private final CheckedBiFunction<AirbyteStreamNameNamespacePair, ConfiguredAirbyteCatalog, RecordBufferImplementation, Exception> onCreateBuffer;
+  private final CheckedBiConsumer<AirbyteStreamNameNamespacePair, RecordBufferImplementation, Exception> onStreamFlush;
+  private VoidCallable onFlushEventHook;
+
+  private Map<AirbyteStreamNameNamespacePair, RecordBufferImplementation> allBuffers = new HashMap<>();
+  private long totalBufferSizeInBytes;
+  private final ConfiguredAirbyteCatalog catalog;
+
+  public SerializedRecordBufferingStrategy(final CheckedBiFunction<AirbyteStreamNameNamespacePair, ConfiguredAirbyteCatalog, RecordBufferImplementation, Exception> onCreateBuffer,
+                                           final ConfiguredAirbyteCatalog catalog,
+                                           final CheckedBiConsumer<AirbyteStreamNameNamespacePair, RecordBufferImplementation, Exception> onStreamFlush) {
+    this.onCreateBuffer = onCreateBuffer;
+    this.catalog = catalog;
+    this.onStreamFlush = onStreamFlush;
+    this.totalBufferSizeInBytes = 0;
+    this.onFlushEventHook = null;
+  }
+
+  @Override
+  public void registerFlushEventHook(final VoidCallable onFlushEventHook) {
+    this.onFlushEventHook = onFlushEventHook;
+  }
+
+  @Override
+  public void addRecord(final AirbyteStreamNameNamespacePair stream, final AirbyteMessage message) throws Exception {
+
+    final RecordBufferImplementation streamBuffer = allBuffers.computeIfAbsent(stream, k -> {
+      LOGGER.info("Starting a new buffer for stream {} (current state: {} in {} buffers)",
+          stream.getName(),
+          FileUtils.byteCountToDisplaySize(totalBufferSizeInBytes),
+          allBuffers.size());
+      try {
+        return onCreateBuffer.apply(stream, catalog);
+      } catch (final Exception e) {
+        LOGGER.error("Failed to create a new buffer for stream {}", stream.getName(), e);
+        throw new RuntimeException(e);
+      }
+    });
+    if (streamBuffer != null) {
+      final Long actualMessageSizeInBytes = streamBuffer.accept(message.getRecord());
+      totalBufferSizeInBytes += actualMessageSizeInBytes;
+      if (totalBufferSizeInBytes > streamBuffer.getMaxTotalBufferSizeInBytes()
```

Comment:
is there a reason not to `flushAll()` if `totalBufferSizeInBytes == streamBuffer.getMaxTotalBufferSizeInBytes()`?